### PR TITLE
bump(llama-cpp): b8708 → b8816

### PR DESCRIPTION
**Pull request by mhamann**
## Summary

Second bump in the 2–3 bump sequence anticipated by the `bump-llama-cpp-gemma4` design doc (Decision 2). Propagates `b8816` through all five locations per the rules in `.claude/rules/llama_cpp_bindings.md`.

**Header diff: zero binding-layer risk.** `include/llama.h` is byte-identical between b8708 and b8816. `ggml/include/ggml.h` only gains documentation comments + a new `ggml_rope_multi` declaration that we don't call.

## Pre-merge checks

- [x] packages/llamafarm-llama tests — 17 passed
- [x] cli/internal/llamabinary tests — pass
- [x] Workflow-dispatch smoke test at b8816 succeeded: run 24531256375. Artifact uploaded, same layout as b8708.
- [x] Pi ARM64 runtime smoke test (see Empirical results below)

## Empirical results: Pi 5 Cortex-A76 / Linux ARM64

Swapped the built artifact onto a live Raspberry Pi 5 and replayed three models at temperature=0 with top-5 logprobs for the prompt "The capital of France is":

| Model | b8708 | b8816 | Ref (Mac CPU) |
|---|---|---|---|
| Qwen 2.5 0.5B Q8_0 | Paris -0.918 | Paris -0.918 (unchanged) | Paris -0.858 |
| Gemma 3 1B Q8_0 base | **SEGFAULT** | "France" -0.582 (no crash) | Paris -0.103 |
| Gemma 3 270M Q8_0 fine-tune | "bel" -0.602 (garbage) | "bel" -0.602 (byte-identical) | Paris -0.003 |

Takeaways:
1. No regressions — Qwen is byte-identical across versions.
2. Real fix — Gemma 3 1B no longer segfaults on this ARM build.
3. The 270M Gemma 3 numerics bug is not fixed in b8816. Logits byte-identical to b8708. Filing upstream separately.

## Follow-up

- Upstream issue for the 270M Gemma 3 ARM CPU numerics bug (will link when filed)
- Next bump (#3) when upstream picks up the fix

### CODE REVIEW

**Comment by mhamann** [comment]
Upstream llama.cpp issue filed: https://github.com/ggml-org/llama.cpp/issues/22011 — full reproduction, environment, and diagnostic logs captured there.

**Comment by mhamann** [comment]
**Bisection update — b7694 produces correct output on the same Pi.**

To confirm that this is a regression rather than \"Gemma 3 was never right on ARM CPU\", I built `llama-cli` from source at llamafarm's previous pin (b7694, the long-lived pre-Gemma-4 pin) for linux/arm64/Debian 12 and ran it against the same v2 Q8_0 GGUF on the same Pi 5 / Cortex-A76:

| Build | Output |
|---|---|
| **b7694** | ✓ `"Paris"` (158 t/s prompt, 48 t/s generation) |
| b8708 | ✗ `"bel"` garbage |
| b8816 | ✗ `"bel"` garbage (byte-identical to b8708) |

So there's a real regression somewhere in the ~1000 commits between b7694 and b8708. Doesn't block merging this bump (#819 is strictly an improvement — one segfault fixed, no regressions) but it does tell us bump #3 won't help either unless the upstream fix lands first.

Posted the bisection window as a follow-up comment on ggml-org/llama.cpp#22011.

**Comment by mhamann** [comment]
**Update: the Pi ARM64 garbage-output bug is a toolchain issue in our CI, not a source-code regression.**

I tested two builds of the exact same `b8708` source tag against the same Pi 5 / Cortex-A76:

| Build container | Compiler | Output |
|---|---|---|
| `ubuntu:24.04` arm64 (matches `build-llama.yml`) | GCC 13.3.0 | ✗ garbage |
| `debian:12` arm64 | GCC 12.2.0 | ✓ correct |

Same cmake flags, same source, same test. The GCC 13.3 ARM64 build produces bad output on our 270M Gemma 3 fine-tune; the GCC 12.2 build does not. Qwen 2.5 is unaffected in both builds (so only one of ggml-cpu's kernels is being miscompiled, and that kernel isn't on Qwen's hot path).

**Concrete fix:** pin the build container / compiler in `.github/workflows/build-llama.yml`. Simplest option is to add an explicit `container:` directive using debian:12 and let ubuntu-24.04-arm host it. Alternative is to install gcc-12 on the Ubuntu runner and export `CC=gcc-12 CXX=g++-12` before cmake.

Can fold that fix into this PR (making it "bump + build-hardening") or split into its own PR — flagging for reviewer preference. Filed the underlying miscompile at [ggml-org/llama.cpp#22011](https://github.com/ggml-org/llama.cpp/issues/22011).

**Comment by mhamann** [comment]
Pushed a commit on this branch that installs `gcc-12`/`g++-12` on the `ubuntu-24.04-arm` runner and compiles with those instead of the default gcc-13.3 — keeps the produced binary ABI-compatible with our Ubuntu 24.04 edge-runtime container, just swaps the compiler.

**End-to-end validation before pushing:** built the exact same b8708 source tag in an `ubuntu:24.04` arm64 container with `gcc-12` installed, copied the resulting `llama-cli` + `*.so` set into our running `arc-edge-runtime-1` container on a Pi 5 / Cortex-A76, and ran the canary:

```
$ llama-cli --model model.Q8_0.gguf --prompt "The capital of France is" --predict 5 --temp 0
> The capital of France is
|- Answer: Paris
[ Prompt: 110.3 t/s | Generation: 45.9 t/s ]
```

vs the same build with stock gcc-13.3 on the same image: ` bel` at -0.602 (the original garbage).

Triggered `build-llama.yml` against this branch for a CI smoke test: [run 24545365762](https://github.com/llama-farm/llamafarm/actions/runs/24545365762). Will wait for it to go green before closing out.

Decision point for you: I bundled the CI fix into this PR since it unblocks the bump's usefulness and it's a ~13-line change. Happy to split into a separate PR if you'd rather keep the "bump should be reviewable in 5 minutes" rule intact — let me know.

**Comment by mhamann** [comment]
**Retracting the CI fix — reverted in 625bd37f.**

Followed up with a more rigorous test after pushing: kept the stock Ubuntu 24.04 / GCC 13.3 binary in place but fixed two bugs in our own `llamafarm-llama` wrapper that were silently producing the bad output I'd been attributing to the compiler. With the wrapper fixed, the GCC 13.3 artifact produces identical output to the GCC 12 one — same `" Paris"` raw completion, same `{"cmd": "takeoff", "altitude": 20}` on the real mission-router v2 GGUF.

The wrapper bugs were:
1. `_apply_chat_template` was passing `NULL` to `llama_chat_apply_template`, which silently falls back to ChatML in b7376+ (instead of the model's embedded template, as the old behavior did). Every non-ChatML model got ChatML delimiters its tokenizer didn't recognize.
2. `create_completion` was tokenizing raw prompts with `add_special=False`, stripping BOS. Decoder-only models (Gemma specifically) degenerate hard without BOS.

Both fixed in #820.

My earlier bisect was the methodological mistake: I tested "good" commits with `llama-cli` (which tokenizes with BOS correctly) and "bad" commits via the Python wrapper HTTP path (which didn't). Two variables changing at once. The divergence I attributed to GCC was actually the wrapper — same runtime behavior across any compiler. Sorry for the noise on #22011 upstream; closed that with an explanation.

PR is now just the b8708→b8816 version bump, which is still independently useful (1B Gemma 3 segfault on ARM → non-crashing under b8816). Reviewable in 5 minutes again.
